src/main/java/de/upb/crypto/math/interfaces/structures/GroupElement.java

```diff
@@ -164,6 +163,30 @@ default GroupElement precomputePow() {
      */
     GroupElement precomputePow(int windowSize);
 
+    /**
+     * Similar to {@link GroupElement#precomputePow()} but precomputes negative powers of this base instead.
+     * Only useful for exponentiations with negative exponents and if you plan on using the
+     * sliding window exponentiation algorithm as wNAF uses the positive powers in any case.
+     * Sliding window, however, cannot assume that the existing positive odd powers are easy to invert and
+     * will use negative powers instead.
+     * @return the same object (for chaining calls)
+     */
+    default GroupElement precomputeNegPow() {
```

Comment:
> I think I don't like this method. What's the high-level promise what will happen if you call it? What would you tell users of this interface when/whether they should call it?
> `precomputePow` is much simpler in this regard - call it and future operations will be faster (but you pay memory and precomputation time).
> 
> Can we just leave this up to the implementation of the `Group` to handle this detail on whether or not negative exponents are precomputed? I would suggest to remove `GroupElement::precomputeNegPow` and instead just have `LazyGroup` somehow decide whether negative exponents are computed during regular `GroupElement::precomputePow` calls or not.

How is that going to work though? The `GroupElement` has no idea whether you will need negative or positive exponents, unless you tell the method whether you will use a negative exponent or not. Which could be an alternative to having two different methods. Then the user would only need to tell us the exponent signum and not have to understand any rest of the exponentiation details. What is actually done could then be decided based on the configured multi-exponentiation algorithm in the `LazyGroup`, i.e. for wNAF we just compute the powers of which we already have the most and for sliding window we select based on exponent signum. 

We run into a problem though if the multi-exponentiation and exponentiation algorithms don't match since then we cannot make a clear decision. With that in mind I don't really like this approach. So I am not sure there are any other options than what I have implemented so far.

Regarding the javadoc, I think the "this makes things faster when using `pow()` with a negative exponent" idea seems reasonable. We can always note that this is not necessarily true and elaborate further down. Then a user without experience in multi-exponentiation algorithms could just abide by the above comment and still be fine, and a more experienced user could still refer to the more exact description. Computing negative powers for wNAF won't really hurt, after all (Shouldn't elements in the wNAF representation be fairly evenly distributed regarding their signum?).

Seems fairly intuitive to me actually: `precomputePow()` for positive exponents and `precomputeNegPow()` for negative exponents. If you have both, call both methods.

I don't think having to write `g.precomputePow().precomputeNegPow()` to get the best possible exponentiation speed later is really a problem. If we were so concerned with having to write a lot, we wouldn't have such long class names. The IDE does most of the writing anyway. Not sure what you mean by saving a factor of two in memory? You are sacrificing memory for later exponentiation speed, are you not?

Reply:
> How is that going to work though? The GroupElement has no idea whether you will need negative or positive exponents

I think we can make reasonable default decisions for that. Like: if inversion is _really_ cheap, don't precompute negative exponents. If inversion takes long / when in doubt, precompute negative exponents.
I'd even be fine with just always precomputing negative exponents, taking the x2 memory hit - my feeling is window sizes larger than 8 don't improve anything noticably, so it's hard to find an application that would run out of memory doing precomputations of window size 8, I think. We could make this an option in the `LazyGroup` whether or not negative exponents are precomputed.

> We run into a problem though if the multi-exponentiation and exponentiation algorithms don't match since then we cannot make a clear decision

When in doubt, just do the precomputation. Memory is not that precious, I think. 

> I don't think having to write g.precomputePow().precomputeNegPow() to get the best possible exponentiation speed later is really a problem

Is there any use-case where the high-level protocol designer would _not_ always just do both precomputations? Like you said, it doesn't hurt for wNAF (probably helps slightly).
The only time I can imagine where I wouldn't do both precomputations would be if (1) I run out of memory with negative exponents, but won't run out of memory without them. Or maybe if I'm being really careful, want to (prematurely) optimize, and (2) know that the current exponentiation choice of the `LazyGroup` doesn't profit at all from negative exponents (i.e. not wNAF and I'll never use negative exponents for that base and never invert the base).
This seems very niche to me. 

tl;dr: I'd advocate for `precomputePow()` to just (usually) implicitly run `precomputeNegPow()`, too (unless we know that inversion in the group is so fast that gains are negligible). If you really want to, I like your suggestion to add an optional parameter to `precomputePow()` that indicates whether or not to precompute negative exponents (default: "do negative exponents" or "let `GroupImpl` decide (reasonable default through inversion cost + whatever other factors)").

Reply:
Sidenote: Okay, I guess technically, it _may_ actually hurt for wNAF to use precomputed negative exponents iff memory access is more costly than group inversion. But since we're unable to optimize that deeply (inversion would always allocate new Java objects, which I guess will be written to memory anyway), I think we can assume that memory access is always less costly than inversion.

Reply:
> (2) know that the current exponentiation choice of the LazyGroup doesn't profit at all from negative exponents (i.e. not wNAF and I'll never use negative exponents for that base and never invert the base).
This seems very niche to me.

This was kind of my issue with it since computing both could potentially double the precomputation cost if inv and op cost the same. If you think that is unlikely to occur, I suppose always doing both is reasonable.